forc-plugins/forc-doc/src/doc.rs

```diff
@@ -253,6 +253,44 @@ impl ModuleInfo {
             .map(|file_path_str| file_path_str.to_string())
             .ok_or_else(|| anyhow::anyhow!("There will always be at least the item name"))
     }
+    /// Compares the current `module_info` to the next `module_info` to determine how many directories to go back to make
+    /// the next file path valid, and returns that path as a `String`.
+    ///
+    /// Example:
+    /// ```
+    /// // number of dirs:               [match][    2    ][    1    ]
+    /// current_location = "project_root/module/submodule1/submodule2/struct.Name.html";
+    /// next_location    =              "module/other_submodule/enum.Name.html";
+    /// result           =               "../../other_submodule/enum.Name.html";
+    /// ```
+    /// In this case the first module to match is "module", so we have no need to go back further than that.
+    pub(crate) fn file_path_from_location(
+        &self,
+        file_name: &str,
+        current_module_info: &ModuleInfo,
+    ) -> Result<String> {
+        let mut mid = 0; // the index to split the module_info from call_path at
+        let mut offset = 0; // the number of directories to go back
+        let mut next_location_iter = self.0.iter().rev().enumerate().peekable();
+        while let Some((index, prefix)) = next_location_iter.peek() {
+            for (count, module) in current_module_info.0.iter().rev().enumerate() {
+                if module == *prefix {
+                    offset = count;
+                    mid = self.0.len() - index;
+                    break;
+                }
+            }
+            next_location_iter.next();
+        }
+        let mut new_path = (0..offset).map(|_| "../").collect::<String>();
+        write!(
+            new_path,
+            "{}/{}",
+            self.0.split_at(mid).1.join("/"),
+            file_name
+        )?;
+        Ok(new_path)
```

Comment:
I guess technically it's both? 🤔 

It's a path to a file you open in your browser. It does technically end in `.html` though.

Reply:
The reason why the distinction is important is that paths work differently on different OSes, whereas URLs are the same. E.g. paths on Windows use the `\` separator (so the `join("/")` would be invalid), while Unix uses `/`. In URLs, both use `/`. It looks like this function is only used for URLs though? If that is the case, it might be clearer to call this `file_url_from_location` or similar. Not super important as we don't test for Windows support atm anyways.

Reply:
I see what you mean. These are definitely paths, they just display in a browser. You can see it where the link would be in the screenshots `file:///home/usr/...`. In that case if we need to handle Windows users as well, I have a solution in mind and I can implement it after this goes in. 

BTW, I'm waiting on #3744 to go through so I can make sure that links to private items don't get created. @emilyaherbert says **soon**TM

Reply:
> In that case if we need to handle Windows users as well, I have a solution in mind and I can implement it after this goes in.

Yeah it's not important enough for us to block this or anything! But in general, it's good to keep in mind that when working with paths, we should prefer Rust's handy `Path` and `PathBuf` types, which have loads of useful methods and take care of using the correct separator for us.

Reply:
Will follow up with #4070